### DISCUSSION TITLE
@QueryParam with null, how to deal with it?

### DISCUSSION THREAD

**Discussion by natros** [DominoKit/domino-rest#65] [Q&A]
I have this get method

```
@GET
List<TemplateDto> getTemplates(
    @QueryParam("categoriaId") @Nullable Integer categoriaId,
    @QueryParam("tipo") List<TipoTemplate> tipos,
    @QueryParam("hidden") boolean hidden,
    @QueryParam("zero") boolean zero,
    @QueryParam("ts") long ts);
```

that generates

```
  public ServerRequest<Void, List<TemplateDto>> getTemplates(Integer categoriaId,
      List<TipoTemplate> tipos, boolean hidden, boolean zero, long ts) {
    TemplatesResource_getTemplates instance = new TemplatesResource_getTemplates();
    instance.setQueryParameter("categoriaId", instance.emptyOrStringValue(() -> categoriaId));
    instance.setQueryParameter("tipo", instance.emptyOrStringValue(() -> tipos));
    instance.setQueryParameter("hidden", instance.emptyOrStringValue(() -> hidden));
    instance.setQueryParameter("zero", instance.emptyOrStringValue(() -> zero));
    instance.setQueryParameter("ts", instance.emptyOrStringValue(() -> ts));
    return instance;
  }
```

and creates URL
`http://localhost:7021/docs/api/templates?categoriaId=&tipo=[PRINCIPAL,%20SECUNDARIO]&hidden=false&zero=false&ts=1615035551000`

The problem is that RestEasy cannot convert '' to Integer

`RESTEASY003870: Unable to extract parameter from http request: javax.ws.rs.QueryParam("categoriaId") value is '' for public abstract java.util.List pt.ipb.gdoc2.shared.api.TemplatesResource.getTemplates(java.lang.Integer,java.util.List,boolean,boolean,long)`

RestyGWT generates a different URL that omits null parameters
`http://localhost:7021/docs/api/templates?zero=false&tipo=PRINCIPAL&tipo=SECUNDARIO&hidden=false&ts=1615036682000`

Do you have any advice in this case?

**Comment by vegegoku** [comment]
Do wr have any specifications in this regard with jax-rs or should we implement our own solution?

**Reply by natros** [reply]
I'm my opinion, it should behave like restygwt to make migration easy. That means if the parameter is null it should not be included. But I don't know the impact this will have on users in the community who are using domino-rest.

**Reply by vegegoku** [reply]
We can make this an option.. WDYT?

**Reply by vegegoku** [reply]
ok i found this https://docs.oracle.com/cd/E19798-01/821-1841/6nmq2cp1v/index.html

> If @DefaultValue is not used in conjunction with @QueryParam, and the query parameter is not present in the request, the value will be an empty collection for List, Set, or SortedSet; null for other object types; and the default for primitive types.

**Reply by natros** [reply]
So, this means that domino-rest does not comply with the specification.

**Reply by vegegoku** [reply]
We used to emit `null` for this case, but this also gave problems to some users too.

What I am thinking of is a global configuration that sets the default behavior for this case, which can be overridden for a specific service or resource method.

plus adding support for `@DefaultValue` for query parameters.

**Reply by natros** [reply]
BTW, RestEasy cannot handle `type=[ENUM1, ENUM2]` for lists. RestyGwt translates to `type=ENUM1&type=ENUM2`

`@QueryParam("type") List<TemplateTypes> types`

where `TemplateTypes` is an `enum { ENUM1, ENUM2}`

Do you want me to create issues for both cases?

**Reply by vegegoku** [reply]
Spring on the other hand works with the comma-separated style, so this also should be a configuration, Yes please an issue will be awesome and thank you for the feedback :+1:

**Reply by natros** [reply]
Done #66 and #67 

About spring. I'm not spring user, but after reading [this](https://stackoverflow.com/questions/9768509/can-spring-mvc-handle-multivalue-query-parameter), it seems that spring supports `url?phone=123&phone=345` and `url?phone=123,345` but I did not find the pattern `url?phone=[123,345]`

**Reply by natros** [reply]
If you go this route to set up the configuration, maybe the more appropriate place is the annotation `@RequestFactory`

If you don't want to have too many configuration parameters in @RequestFactory, you can always implement something like `
@RequestFactory(config = MyRequestConf.class)` where `MyRequestConf` could be described as

```java
@RequestFactoryConfig (option1 = ..., option2 = ..., optionN = ...)
public interface MyRequestConf {}
```

**Reply by vegegoku** [reply]
> If you go this route to set up the configuration, maybe the more appropriate place is the annotation `@RequestFactory`
> 
> If you don't want to have too many configuration parameters in @RequestFactory, you can always implement something like ` @RequestFactory(config = MyRequestConf.class)` where `MyRequestConf` could be described as
> 
> ```java
> @RequestFactoryConfig (option1 = ..., option2 = ..., optionN = ...)
> public interface MyRequestConf {}
> ```

I think we are going to utilize the https://github.com/DominoKit/domino-rest/blob/development/domino-rest-client/src/main/java/org/dominokit/rest/DominoRestConfig.java for the global configurations 

for overriding the behavior for the specific interface we will go with new annotations.
